File: src/main/java/algo/interval/FindRightInterval_436.java
package algo.interval;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;

public class FindRightInterval_436 {
  public static int[] findRightInterval(int[][] intervals) {
    var len = intervals.length;
    var startPosition = new HashMap<Integer, Integer>();

    // record position of each interval start
    var maxStart = 0;
    for (var i = 0; i < len; i++) {
      startPosition.put(intervals[i][0], i);
      maxStart = Math.max(maxStart, intervals[i][0]);
    }

    var result = new int[len];
    Arrays.fill(result, -1);
    for (var i = 0; i < len; i++) {
      var interval = intervals[i];
      // find the smallest start
      for (var j = interval[1]; j <= maxStart; j++) {
        if (startPosition.containsKey(j)) {
          result[i] = startPosition.get(j);
          break;
        }
      }
    }
    return result;
  }

  public static void main(String[] args) {
    int[] right;
//    right = findRightInterval(new int[][]{{1,2}});
//    right = findRightInterval(new int[][]{{3,4}, {2,3}, {1,2}});
//    right = findRightInterval(new int[][]{{1,4}, {2,3}, {3,4}});
    right = findRightInterval(new int[][]{{1,1}, {3,4}});
    for (var n : right) {
      System.out.println(n);
    }
  }
}

File: src/main/kotlin/algo/interval/FindRightInterval_436.kt
package algo.interval

class KtFindRightInterval_436 {
    fun findRightInterval(intervals: Array<IntArray>): IntArray {
        // position of each interval start
        var maxStart = 0
        val startPosition: MutableMap<Int, Int> = mutableMapOf()
        intervals.indices.forEach {
            startPosition[intervals[it][0]] = it
            maxStart = maxStart.coerceAtLeast(intervals[it][0])
        }

        val result = IntArray(intervals.size) { -1 }
        intervals.indices.forEach {
            val current = intervals[it]
            // find the smallest start in the map
            for (i in current[1]..maxStart) {
                if (startPosition.containsKey(i)) {
                    result[it] = startPosition[i]!!
                    break
                }
            }
        }
        return result
    }
}